# Data Cleaning
Since data scientist spends more than half or their times cleaning data, in this projet will learn the data analysis cleaning tasks:
*  Cleaning columns names
* Extract and convert numeric values form strings values
* Extract string data
* Handle missing values

We will work with a dataset laptops.csv a dataset about of 1300 computers.

### 1. Reading csv file with encoding
Nowadayas the predominent encoding is UTF-8 which the default encoding of Python. It exists others encoding Latin-1 (also known as ISO-8895-1) and Windows-1251.
### Instructions
* Import the pandas library
* Use the pandas.read_csv() function to read the laptops.csv file into a dataframe laptops.
  * Specify the encoding using the string "Latin-1".
  * If that doesn't work, try using the string "Windows-1251".
* Use the DataFrame.info() method to display information about the laptops dataframe.

In [1]:
import pandas as pd
laptops = pd.read_csv("laptops.csv", encoding="Latin-1")
laptops.head()

Manufacturer   Model Name   Category Screen Size  \
0        Apple  MacBook Pro  Ultrabook       13.3"   
1        Apple  Macbook Air  Ultrabook       13.3"   
2           HP       250 G6   Notebook       15.6"   
3        Apple  MacBook Pro  Ultrabook       15.4"   
4        Apple  MacBook Pro  Ultrabook       13.3"   

                               Screen                         CPU   RAM  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

               Storage                           GPU Operating System  \
0            128GB SSD  Intel Iris Plus Graphics 640            macOS   
1  128GB Flash Storage        Intel HD Graphics 6000            macOS   
2            256GB SSD         Intel HD Graphics 620            No OS   
3            512GB SSD            AMD Radeon Pro 455            macOS   
4            256GB SSD  Intel Iris Plus Graphics 650            macOS   

  Operating System Version  Weight Price (Euros)  
0                      NaN  1.37kg       1339,69  
1                      NaN  1.34kg        898,94  
2                      NaN  1.86kg        575,00  
3                      NaN  1.83kg       2537,45  
4                      NaN  1.37kg       1803,60

In [2]:
laptops.info(verbose = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
Manufacturer                1303 non-null object
Model Name                  1303 non-null object
Category                    1303 non-null object
Screen Size                 1303 non-null object
Screen                      1303 non-null object
CPU                         1303 non-null object
RAM                         1303 non-null object
 Storage                    1303 non-null object
GPU                         1303 non-null object
Operating System            1303 non-null object
Operating System Version    1133 non-null object
Weight                      1303 non-null object
Price (Euros)               1303 non-null object
dtypes: object(13)
memory usage: 132.4+ KB


In [3]:
laptops.describe()

Manufacturer Model Name  Category Screen Size             Screen  \
count          1303       1303      1303        1303               1303   
unique           19        618         6          18                 40   
top          Lenovo     XPS 13  Notebook       15.6"  Full HD 1920x1080   
freq            297         30       727         665                507   

                               CPU   RAM    Storage                    GPU  \
count                         1303  1303       1303                   1303   
unique                         118     9         38                    110   
top     Intel Core i5 7200U 2.5GHz   8GB  256GB SSD  Intel HD Graphics 620   
freq                           190   619        412                    281   

       Operating System Operating System Version Weight Price (Euros)  
count              1303                     1133   1303          1303  
unique                7                        4    179           791  
top             Windows                       10  2.2kg       1799,00  
freq               1125                     1072    121            14

### 2. Cleaning columns names
The column labels have a variety of upper and lowercase letters, as well as spaces and parentheses. Because we represent column labels as strings, this isn't inherently a problem, but you'll find that uniform column labels will be easier to work with. We'll clean our column labels by:

Removing any whitespace from the start and end of the labels - if you look closely you'll notice that the " Storage" column label has a space in front of it. These quirks with column labels are often hard to spot, so removing them at the start will save you pain in the long run.
Replacing spaces with underscores and remove special characters - this will make things consistent, and also allows for the use of dot accessors for those who prefer that.
Make all labels lowercase - this is good for consistency, and means you'll never have to remember what is capitalized and how.
Shorten any long column names - this helps to keep your code easier to read, especially when you are using method chaining.

### Instructions
* Define a function, which accepts a string argument, and:
  * Removes any whitespace from the start and end of the string.
  * Replaces the substring Operating System with the abbreviation os.
  * Replaces all spaces with underscores.
  * Removes parentheses from the string.
  * Makes the entire string lowercase.
  * Returns the modified string.
* Use a loop to apply the function to each item in the DataFrame.column attribute for the laptops dataframe, assigning the result back to the the DataFrame.columns attribute.

In [4]:
# cleanig the dataset columns labels, remove white spaces, transform labels to lowercase,
# replace space with underscore for dot label accessor
columns = laptops.columns
print(columns)

Index(['Manufacturer', 'Model Name', 'Category', 'Screen Size', 'Screen',
       'CPU', 'RAM', ' Storage', 'GPU', 'Operating System',
       'Operating System Version', 'Weight', 'Price (Euros)'],
      dtype='object')


In [5]:
# clean columns with a function
def clean_column(colname):
    """
    This function take as a paramter a string and clean it removing
    whitespaces, replace them with underscore, replace "("" & ")" with
    nothing
    """
    colname = colname.strip().replace("Operating System", "os")
    colname = colname.replace(' ', '_').replace("(", "").replace(")", "")
    colname = colname.lower()
    return colname

# iteration the DataFrame columns names
cleaned_columns = []
for item in columns:
    cleaned_item = clean_column(item)
    cleaned_columns.append(cleaned_item)

# assign the DataFrame columns names to the cleaned_columns
laptops.columns = cleaned_columns
print(laptops.columns)

Index(['manufacturer', 'model_name', 'category', 'screen_size', 'screen',
       'cpu', 'ram', 'storage', 'gpu', 'os', 'os_version', 'weight',
       'price_euros'],
      dtype='object')


In [6]:
laptops.info(verbose=True) # all columns are of object type indicating that they all string

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 13 columns):
manufacturer    1303 non-null object
model_name      1303 non-null object
category        1303 non-null object
screen_size     1303 non-null object
screen          1303 non-null object
cpu             1303 non-null object
ram             1303 non-null object
storage         1303 non-null object
gpu             1303 non-null object
os              1303 non-null object
os_version      1133 non-null object
weight          1303 non-null object
price_euros     1303 non-null object
dtypes: object(13)
memory usage: 132.4+ KB


### 3. Converting string columns to numeric

In [7]:
laptops.head()

manufacturer   model_name   category screen_size  \
0        Apple  MacBook Pro  Ultrabook       13.3"   
1        Apple  Macbook Air  Ultrabook       13.3"   
2           HP       250 G6   Notebook       15.6"   
3        Apple  MacBook Pro  Ultrabook       15.4"   
4        Apple  MacBook Pro  Ultrabook       13.3"   

                               screen                         cpu   ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

               storage                           gpu     os os_version  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS        NaN   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS        NaN   
2            256GB SSD         Intel HD Graphics 620  No OS        NaN   
3            512GB SSD            AMD Radeon Pro 455  macOS        NaN   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS        NaN   

   weight price_euros  
0  1.37kg     1339,69  
1  1.34kg      898,94  
2  1.86kg      575,00  
3  1.83kg     2537,45  
4  1.37kg     1803,60

### Converting text to numeric workflow
Whenever we're converting text to numeric data, we can follow this data cleaning workflow:
<img src="clean_column.png"/>

In [8]:
# Explore the data in the column
# Use Series.unique() to explore unique values of the columns
print(laptops["screen_size"].dtype)
print(laptops["screen_size"].unique())

object
['13.3"' '15.6"' '15.4"' '14.0"' '12.0"' '11.6"' '17.3"' '10.1"' '13.5"'
 '12.5"' '13.0"' '18.4"' '13.9"' '12.3"' '17.0"' '15.0"' '14.1"' '11.3"']


Our next stage is to **identify patterns and special cases**. We can see that all values in this column follow the same pattern - series of digit and period characters, followed by a quote character. There are no special cases - every value matches the same pattern. We can also observe that we will need to convert the column to a float dtype, as the int dtype won't be able to store the decimal values.

The next stage is to **remove the non-digit characters**. The pandas library contains dozens of vectorized string methods, most of which are available using the Series.str accessor. In this case, we can use the Series.str.replace() method, which is a vectorized version of the Python str.replace() method we used in the previous screen to remove all the quote characters:

In [9]:
laptops["screen_size"] = laptops["screen_size"].str.replace('"','') # replace " with nothing
print(laptops["screen_size"].unique())

['13.3' '15.6' '15.4' '14.0' '12.0' '11.6' '17.3' '10.1' '13.5' '12.5'
 '13.0' '18.4' '13.9' '12.3' '17.0' '15.0' '14.1' '11.3']


Now we've removed the non-digit characters, we can convert (or cast) the column to a numeric dtype. To do this, we use the [Series.astype() method](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.astype.html). We can use either int or float as the parameter for the method to convert the column to the respective type:

In [10]:
laptops["screen_size"] = laptops["screen_size"].astype(float)
print(laptops["screen_size"].dtype)
print(laptops["screen_size"].unique())

float64
[13.3 15.6 15.4 14.  12.  11.6 17.3 10.1 13.5 12.5 13.  18.4 13.9 12.3
 17.  15.  14.1 11.3]


Our **final step is to rename the column**. This is an optional step, and can be useful if the non-digit values contained information that helps us understand the data. In our case, the quote characters actually denoted that the screen size was in inches. We can use the [DataFrame.rename() method]() to rename specific axis labels using a dictionary with the keys as the old label name, and the values as the new label name.

In [11]:
laptops.rename({"screen_size": "screen_size_inches"}, axis=1, inplace=True)
print(laptops.dtypes)

manufacturer           object
model_name             object
category               object
screen_size_inches    float64
screen                 object
cpu                    object
ram                    object
storage                object
gpu                    object
os                     object
os_version             object
weight                 object
price_euros            object
dtype: object


### Intructions
* Use the Series.str.replace() method to remove the substring GB from the ram column.
* Use the Series.astype() method to change the ram column to an integer dtype.
* Because the GB characters contained useful information about the units (gigabytes) of the laptop's ram, use the DataFrame.rename() method to rename the column from ram to ram_gb.
* Use the DataFrame.dtypes attribute to get a list of the column names and types from the laptops dataframe, and assign the result to dtypes.
* After you have run your code, use the variable inspector to view the dtypes variable to see the results of your code.

In [12]:
# exploring the ram columns
print(laptops['ram'].dtype)
print(laptops['ram'].unique())

object
['8GB' '16GB' '4GB' '2GB' '12GB' '6GB' '32GB' '24GB' '64GB']


In [13]:
# identify a pattern and special cases
# here remove GB string and convert value to int type with astype() method
laptops["ram"] = laptops["ram"].str.replace("GB","").astype(int)
print(laptops["ram"].dtype)
print(laptops["ram"].unique())

int32
[ 8 16  4  2 12  6 32 24 64]


In [14]:
# rename columns name
laptops.rename({"ram":"ram_gb"}, axis=1, inplace=True)
laptops.head()

manufacturer   model_name   category  screen_size_inches  \
0        Apple  MacBook Pro  Ultrabook                13.3   
1        Apple  Macbook Air  Ultrabook                13.3   
2           HP       250 G6   Notebook                15.6   
3        Apple  MacBook Pro  Ultrabook                15.4   
4        Apple  MacBook Pro  Ultrabook                13.3   

                               screen                         cpu  ram_gb  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz       8   
1                            1440x900        Intel Core i5 1.8GHz       8   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz       8   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz      16   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz       8   

               storage                           gpu     os os_version  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS        NaN   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS        NaN   
2            256GB SSD         Intel HD Graphics 620  No OS        NaN   
3            512GB SSD            AMD Radeon Pro 455  macOS        NaN   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS        NaN   

   weight price_euros  
0  1.37kg     1339,69  
1  1.34kg      898,94  
2  1.86kg      575,00  
3  1.83kg     2537,45  
4  1.37kg     1803,60

In [15]:
dtypes = laptops.dtypes
dtypes

manufacturer           object
model_name             object
category               object
screen_size_inches    float64
screen                 object
cpu                    object
ram_gb                  int32
storage                object
gpu                    object
os                     object
os_version             object
weight                 object
price_euros            object
dtype: object

### 4. Practicing, converting string columns to numeric 
### Instructions
* Clean the weight column by:
  - Removing all non-digit characters.
  - Casting the column to the appropriate numeric type
  - Renaming the column to weight_kg.
* Clean the price_euros column by:
  - Replacing the decimal commas with decimal points.
  - Casting the column to the appropriate numeric type.
* Use the Series.describe() method to generate some descriptive statistics for each column:
  - Assign the results for the weight_kg column to weight_describe.
  - Assign the results for the price_euros column to price_describe

In [16]:
print(laptops["weight"].dtype)
print(laptops["weight"].unique())

object
['1.37kg' '1.34kg' '1.86kg' '1.83kg' '2.1kg' '2.04kg' '1.3kg' '1.6kg'
 '2.2kg' '0.92kg' '1.22kg' '0.98kg' '2.5kg' '1.62kg' '1.91kg' '2.3kg'
 '1.35kg' '1.88kg' '1.89kg' '1.65kg' '2.71kg' '1.2kg' '1.44kg' '2.8kg'
 '2kg' '2.65kg' '2.77kg' '3.2kg' '0.69kg' '1.49kg' '2.4kg' '2.13kg'
 '2.43kg' '1.7kg' '1.4kg' '1.8kg' '1.9kg' '3kg' '1.252kg' '2.7kg' '2.02kg'
 '1.63kg' '1.96kg' '1.21kg' '2.45kg' '1.25kg' '1.5kg' '2.62kg' '1.38kg'
 '1.58kg' '1.85kg' '1.23kg' '1.26kg' '2.16kg' '2.36kg' '2.05kg' '1.32kg'
 '1.75kg' '0.97kg' '2.9kg' '2.56kg' '1.48kg' '1.74kg' '1.1kg' '1.56kg'
 '2.03kg' '1.05kg' '4.4kg' '1.90kg' '1.29kg' '2.0kg' '1.95kg' '2.06kg'
 '1.12kg' '1.42kg' '3.49kg' '3.35kg' '2.23kg' '4.42kg' '2.69kg' '2.37kg'
 '4.7kg' '3.6kg' '2.08kg' '4.3kg' '1.68kg' '1.41kg' '4.14kg' '2.18kg'
 '2.24kg' '2.67kg' '2.14kg' '1.36kg' '2.25kg' '2.15kg' '2.19kg' '2.54kg'
 '3.42kg' '1.28kg' '2.33kg' '1.45kg' '2.79kg' '1.84kg' '2.6kg' '2.26kg'
 '3.25kg' '1.59kg' '1.13kg' '1.78kg' '1.10kg' '1.15kg' '1.27kg' 

In [17]:
laptops["weight"] = laptops["weight"].str.replace("kg","").str.replace('s',"").astype(float)
# rename the weight column to weight_kg
laptops.rename({"weight":"weight_kg"}, axis=1, inplace=True)
laptops.head()

manufacturer   model_name   category  screen_size_inches  \
0        Apple  MacBook Pro  Ultrabook                13.3   
1        Apple  Macbook Air  Ultrabook                13.3   
2           HP       250 G6   Notebook                15.6   
3        Apple  MacBook Pro  Ultrabook                15.4   
4        Apple  MacBook Pro  Ultrabook                13.3   

                               screen                         cpu  ram_gb  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz       8   
1                            1440x900        Intel Core i5 1.8GHz       8   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz       8   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz      16   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz       8   

               storage                           gpu     os os_version  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS        NaN   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS        NaN   
2            256GB SSD         Intel HD Graphics 620  No OS        NaN   
3            512GB SSD            AMD Radeon Pro 455  macOS        NaN   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS        NaN   

   weight_kg price_euros  
0       1.37     1339,69  
1       1.34      898,94  
2       1.86      575,00  
3       1.83     2537,45  
4       1.37     1803,60

In [18]:
laptops["price_euros"] = laptops["price_euros"].str.replace(",", ".")
laptops["price_euros"].astype(float)
print(laptops["price_euros"].head())

0    1339.69
1     898.94
2     575.00
3    2537.45
4    1803.60
Name: price_euros, dtype: object


In [19]:
weight_describe = laptops["weight_kg"].describe()
weight_describe

count    1303.000000
mean        2.038734
std         0.665475
min         0.690000
25%         1.500000
50%         2.040000
75%         2.300000
max         4.700000
Name: weight_kg, dtype: float64

In [20]:
price_describe = laptops["price_euros"].describe()
price_describe

count        1303
unique        791
top       1799.00
freq           14
Name: price_euros, dtype: object

## Extractiong values from the start of strings

In [21]:
laptops["gpu"].head(10).str.split(expand = True)  # expand = True return a DataFrame , n = limit the number of split

0        1         2         3     4
0   Intel     Iris      Plus  Graphics   640
1   Intel       HD  Graphics      6000  None
2   Intel       HD  Graphics       620  None
3     AMD   Radeon       Pro       455  None
4   Intel     Iris      Plus  Graphics   650
5     AMD   Radeon        R5      None  None
6   Intel     Iris       Pro  Graphics  None
7   Intel       HD  Graphics      6000  None
8  Nvidia  GeForce     MX150      None  None
9   Intel      UHD  Graphics       620  None

In [22]:
laptops["gpu"].head(10).str.split(n=1, expand = True)

0                       1
0   Intel  Iris Plus Graphics 640
1   Intel        HD Graphics 6000
2   Intel         HD Graphics 620
3     AMD          Radeon Pro 455
4   Intel  Iris Plus Graphics 650
5     AMD               Radeon R5
6   Intel       Iris Pro Graphics
7   Intel        HD Graphics 6000
8  Nvidia           GeForce MX150
9   Intel        UHD Graphics 620

In [23]:
laptops["gpu"].head(10).str.split(expand = True).iloc[:, 0]

0     Intel
1     Intel
2     Intel
3       AMD
4     Intel
5       AMD
6     Intel
7     Intel
8    Nvidia
9     Intel
Name: 0, dtype: object

In [24]:
laptops["cpu"].head(10)

0          Intel Core i5 2.3GHz
1          Intel Core i5 1.8GHz
2    Intel Core i5 7200U 2.5GHz
3          Intel Core i7 2.7GHz
4          Intel Core i5 3.1GHz
5       AMD A9-Series 9420 3GHz
6          Intel Core i7 2.2GHz
7          Intel Core i5 1.8GHz
8    Intel Core i7 8550U 1.8GHz
9    Intel Core i5 8250U 1.6GHz
Name: cpu, dtype: object

In [25]:
laptops["cpu"].head(10).str.split(expand = True)

0          1     2       3       4
0  Intel       Core    i5  2.3GHz    None
1  Intel       Core    i5  1.8GHz    None
2  Intel       Core    i5   7200U  2.5GHz
3  Intel       Core    i7  2.7GHz    None
4  Intel       Core    i5  3.1GHz    None
5    AMD  A9-Series  9420    3GHz    None
6  Intel       Core    i7  2.2GHz    None
7  Intel       Core    i5  1.8GHz    None
8  Intel       Core    i7   8550U  1.8GHz
9  Intel       Core    i5   8250U  1.6GHz

In [26]:
laptops["cpu"].head(10).str.split(expand = True).iloc[:, 0]

0    Intel
1    Intel
2    Intel
3    Intel
4    Intel
5      AMD
6    Intel
7    Intel
8    Intel
9    Intel
Name: 0, dtype: object

## Extractiong values from the end  of strings

In [27]:
unique_values_screen = laptops["screen"].unique()
shape = laptops["screen"].unique().shape

In [28]:
unique_values_screen[:10]

array(['IPS Panel Retina Display 2560x1600', '1440x900',
       'Full HD 1920x1080', 'IPS Panel Retina Display 2880x1800',
       '1366x768', 'IPS Panel Full HD 1920x1080',
       'IPS Panel Retina Display 2304x1440',
       'IPS Panel Full HD / Touchscreen 1920x1080',
       'Full HD / Touchscreen 1920x1080',
       'Touchscreen / Quad HD+ 3200x1800'], dtype=object)

In [29]:
# splitting the the values of the screen to extract the resolution of the screen
laptops["screen"].str.rsplit(n = 1,  expand=True).head(10)

0          1
0  IPS Panel Retina Display  2560x1600
1                  1440x900       None
2                   Full HD  1920x1080
3  IPS Panel Retina Display  2880x1800
4  IPS Panel Retina Display  2560x1600
5                  1366x768       None
6  IPS Panel Retina Display  2880x1800
7                  1440x900       None
8                   Full HD  1920x1080
9         IPS Panel Full HD  1920x1080

In [30]:
laptops["screen"].str.rsplit(n=1, expand=True).iloc[:, 1].head(10)

0    2560x1600
1         None
2    1920x1080
3    2880x1800
4    2560x1600
5         None
6    2880x1800
7         None
8    1920x1080
9    1920x1080
Name: 1, dtype: object

In [31]:
screen_res = laptops["screen"].str.rsplit(n=1, expand=True)

In [32]:
screen_res.columns = ["A", "B"]

In [33]:
screen_res.head()

A          B
0  IPS Panel Retina Display  2560x1600
1                  1440x900       None
2                   Full HD  1920x1080
3  IPS Panel Retina Display  2880x1800
4  IPS Panel Retina Display  2560x1600

In [34]:
laptops["screen_resolution"] = screen_res["B"]
print(laptops["screen_resolution"].unique().shape)
print(laptops["screen_resolution"].unique())

(14,)
['2560x1600' None '1920x1080' '2880x1800' '2304x1440' '3200x1800'
 '1920x1200' '2256x1504' '1366x768' '3840x2160' '2160x1440' '2560x1440'
 '2736x1824' '2400x1600']


## Instructions
* From the cpu column, perform the following steps to extract and convert the processor speed, assigning the results to a new column cpu_speed_ghz:
  * Use Series.str.replace() to remove the substring "GHz" each string.
  * Use Series.str.rsplit() and DataFrame.iloc[] to select the numeric characters from the end of the string.
  * Use Series.astype() to cast the values to the float dtype.

In [35]:
laptops["cpu"].head(10)

0          Intel Core i5 2.3GHz
1          Intel Core i5 1.8GHz
2    Intel Core i5 7200U 2.5GHz
3          Intel Core i7 2.7GHz
4          Intel Core i5 3.1GHz
5       AMD A9-Series 9420 3GHz
6          Intel Core i7 2.2GHz
7          Intel Core i5 1.8GHz
8    Intel Core i7 8550U 1.8GHz
9    Intel Core i5 8250U 1.6GHz
Name: cpu, dtype: object

In [36]:
# remove  GHz in each string
cpu = laptops["cpu"].str.replace("GHz", "").str.rsplit(n=1, expand=True)
cpu.head()

0    1
0        Intel Core i5  2.3
1        Intel Core i5  1.8
2  Intel Core i5 7200U  2.5
3        Intel Core i7  2.7
4        Intel Core i5  3.1

In [37]:
# selection of the column cpu speed
cpu_ghz = cpu.iloc[:,1].astype(float)
cpu_ghz.head()

0    2.3
1    1.8
2    2.5
3    2.7
4    3.1
Name: 1, dtype: float64

In [38]:
laptops["cpu_speed_ghz"] = cpu_ghz

In [39]:
laptops["cpu_speed_ghz"].head()

0    2.3
1    1.8
2    2.5
3    2.7
4    3.1
Name: cpu_speed_ghz, dtype: float64

In [40]:
laptops.head()

manufacturer   model_name   category  screen_size_inches  \
0        Apple  MacBook Pro  Ultrabook                13.3   
1        Apple  Macbook Air  Ultrabook                13.3   
2           HP       250 G6   Notebook                15.6   
3        Apple  MacBook Pro  Ultrabook                15.4   
4        Apple  MacBook Pro  Ultrabook                13.3   

                               screen                         cpu  ram_gb  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz       8   
1                            1440x900        Intel Core i5 1.8GHz       8   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz       8   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz      16   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz       8   

               storage                           gpu     os os_version  \
0            128GB SSD  Intel Iris Plus Graphics 640  macOS        NaN   
1  128GB Flash Storage        Intel HD Graphics 6000  macOS        NaN   
2            256GB SSD         Intel HD Graphics 620  No OS        NaN   
3            512GB SSD            AMD Radeon Pro 455  macOS        NaN   
4            256GB SSD  Intel Iris Plus Graphics 650  macOS        NaN   

   weight_kg price_euros screen_resolution  cpu_speed_ghz  
0       1.37     1339.69         2560x1600            2.3  
1       1.34      898.94              None            1.8  
2       1.86      575.00         1920x1080            2.5  
3       1.83     2537.45         2880x1800            2.7  
4       1.37     1803.60         2560x1600            3.1

## 7. Correction bad values

Here we will use [Series.map method](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html) to correct, bad values of our dataset.

## Instructions
We have created a dictionary for you to use with mapping. Note that we have included both the correct and incorrect spelling of macOS as keys, otherwise we'll end up with null values.

Use the [Series.map() method](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html) with the mapping_dict dictionary to correct the values in the os column.

In [41]:
mapping_dict = {
    'Android': 'Android',
    'Chrome OS': 'Chrome OS',
    'Linux': 'Linux',
    'Mac OS': 'macOS',
    'No OS': 'No OS',
    'Windows': 'Windows',
    'macOS': 'macOS'
}

In [42]:
# correcting bad values
laptops["os"] = laptops["os"].map(mapping_dict)

## Dropping missing values
In previous missions, we've talked briefly about missing values, and how both NumPy and pandas represent these as null values. In pandas null values will be indicated by either NaN or None. Generally the first thing that we want to do is identify which values are missing.

There are two approaches we can use: the [DataFrame.info() method](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html) and the [DataFrame.isnull() method](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isnull.html]. The [DataFrame.info() method](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html) will print information about the dataframe, including the number of non-null values in each column:

In [43]:
laptops.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 15 columns):
manufacturer          1303 non-null object
model_name            1303 non-null object
category              1303 non-null object
screen_size_inches    1303 non-null float64
screen                1303 non-null object
cpu                   1303 non-null object
ram_gb                1303 non-null int32
storage               1303 non-null object
gpu                   1303 non-null object
os                    1303 non-null object
os_version            1133 non-null object
weight_kg             1303 non-null float64
price_euros           1303 non-null object
screen_resolution     989 non-null object
cpu_speed_ghz         1303 non-null float64
dtypes: float64(3), int32(1), object(11)
memory usage: 147.7+ KB


In contrast, DataFrame.isnull() returns a boolean dataframe with True and False indications for every value in the dataframe, and then we can use DataFrame.sum() to give us accounts– using a .sum() method on a boolean array will give us a count of the True values:

In [44]:
# print the count of null values
print(laptops.isnull().sum())

manufacturer            0
model_name              0
category                0
screen_size_inches      0
screen                  0
cpu                     0
ram_gb                  0
storage                 0
gpu                     0
os                      0
os_version            170
weight_kg               0
price_euros             0
screen_resolution     314
cpu_speed_ghz           0
dtype: int64


170 of the os_version are null and 314 columns values of the screen_resolution

We have a few options for how we can handle missing values:
* Remove any rows that have missing values.
* Remove any columns that have missing values.
* Fill the missing values with some other value.
* Leave the missing values as is.

For dropping volume with DataFrame.dropna(axis=0, 1)

## Instructions
* Use DataFrame.dropna() to remove any rows from the laptops dataframe that have null values, assigning the result to laptops_no_null_rows.
* Use DataFrame.dropna() to remove any columns from the laptops dataframe that have null values, assigning the result to laptops_no_null_cols.

In [45]:
laptops_no_null_rows = laptops.dropna(axis = 0, inplace = True)
laptops_no_null_cols = laptops.dropna(axis = 1, inplace = True)

In [46]:
laptops.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 892 entries, 6 to 1299
Data columns (total 15 columns):
manufacturer          892 non-null object
model_name            892 non-null object
category              892 non-null object
screen_size_inches    892 non-null float64
screen                892 non-null object
cpu                   892 non-null object
ram_gb                892 non-null int32
storage               892 non-null object
gpu                   892 non-null object
os                    892 non-null object
os_version            892 non-null object
weight_kg             892 non-null float64
price_euros           892 non-null object
screen_resolution     892 non-null object
cpu_speed_ghz         892 non-null float64
dtypes: float64(3), int32(1), object(11)
memory usage: 108.0+ KB


## 9. Filling missing value


## Instructions

* Use a boolean array to identify rows that have the value No OS for the os column, and then use assignment to assign the value Version Unknown to the os_version column for those rows.
* Use the syntax below to create value_counts_after variable:

  value_counts_after = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()
  

* After you have run your code, use the variable inspector to look at the difference between value_counts_before and value_counts_after.

In [47]:
value_counts_before = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()
laptops.loc[laptops["os"] == "macOS", "os_version"] = "X"
print(value_counts_before)

Series([], Name: os, dtype: int64)


In [48]:
# row with no OS for the os column
laptops.loc[laptops["os"] == 'No OS', "os_version"] = "Version Unknown"
value_counts_after = laptops.loc[laptops["os_version"].isnull(), "os"].value_counts()
print(value_counts_after)

Series([], Name: os, dtype: int64)


## 10. Extracting data storage informations


In [49]:
laptops[["storage"]].head()

storage
6   256GB Flash Storage
8             512GB SSD
9             256GB SSD
13            256GB SSD
16            256GB SSD

In [50]:
laptops["storage"].unique()

array(['256GB Flash Storage', '512GB SSD', '256GB SSD', '128GB SSD',
       '128GB SSD +  1TB HDD', '256GB SSD +  256GB SSD', '1TB HDD',
       '64GB Flash Storage', '500GB HDD', '256GB SSD +  1TB HDD',
       '256GB SSD +  2TB HDD', '1TB Hybrid', '2TB HDD',
       '512GB SSD +  1TB HDD', '1TB SSD', '32GB Flash Storage',
       '256GB SSD +  500GB HDD', '128GB SSD +  2TB HDD',
       '512GB SSD +  512GB SSD', '512GB SSD +  256GB SSD',
       '512GB SSD +  2TB HDD', '32GB SSD',
       '64GB Flash Storage +  1TB HDD', '1GB SSD', '1TB HDD +  1TB HDD',
       '32GB HDD', '1TB SSD +  1TB HDD', '512GB Flash Storage',
       '128GB HDD', '240GB SSD', '8GB SSD', '512GB SSD +  1TB Hybrid',
       '256GB SSD +  1TB Hybrid'], dtype=object)

In [51]:
laptops["storage"].str.split(expand=True).head(10)

0      1        2     3     4     5
6   256GB  Flash  Storage  None  None  None
8   512GB    SSD     None  None  None  None
9   256GB    SSD     None  None  None  None
13  256GB    SSD     None  None  None  None
16  256GB    SSD     None  None  None  None
19  128GB    SSD     None  None  None  None
21  128GB    SSD        +   1TB   HDD  None
23  256GB    SSD     None  None  None  None
24  256GB    SSD     None  None  None  None
27  256GB    SSD     None  None  None  None

In [52]:
laptops["storage"].str.split(expand=True).tail(10)

0       1     2     3     4     5
1275  512GB     SSD  None  None  None  None
1279    1TB     HDD  None  None  None  None
1280    1TB  Hybrid  None  None  None  None
1284  128GB     SSD  None  None  None  None
1285  512GB     SSD  None  None  None  None
1289  512GB     SSD  None  None  None  None
1293    1TB     HDD  None  None  None  None
1294    1TB  Hybrid  None  None  None  None
1298  128GB     SSD  None  None  None  None
1299  512GB     SSD  None  None  None  None

In [53]:
storage_capicity = laptops["storage"].str.split(expand=True).iloc[:, 0]
storage_capicity.head(10)

6     256GB
8     512GB
9     256GB
13    256GB
16    256GB
19    128GB
21    128GB
23    256GB
24    256GB
27    256GB
Name: 0, dtype: object

In [59]:
#laptops["storage_1_capacity_gb"] = storage_capicity.str.replace("GB", "").replace({'1TB': None}).astype(float)
laptops["storage_1_capacity_gb"] = storage_capicity.str.replace('1TB', None).astype(float)

laptops[["storage_1_capacity_gb", "storage"]].head(10)

TypeError: repl must be a string or callable